File: src/gg/obsidian/mobcontrol/MobManager.java
package gg.obsidian.mobcontrol;

import org.bukkit.Chunk;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.*;
import org.bukkit.event.entity.CreatureSpawnEvent;
import org.bukkit.inventory.EntityEquipment;
import org.bukkit.inventory.ItemStack;

import java.util.HashMap;
import java.util.Map;

public class MobManager {
	public final MobControl plugin;

	public MobManager(MobControl instance) {
		plugin = instance;
	}

	public boolean hasCap(Entity entity) {
		return getCap(getCapKey(entity)) >= 0;
	}

	public int getCap(String key) {
		Integer cap = plugin.config.MOB_LIMITS.get(key);
		return (cap != null) ? cap : -1;
	}

	public String getCapKey(Entity entity) {
		StringBuilder key = new StringBuilder(entity.getType().name());

		if (entity instanceof Sheep) {
			key.append(((Sheep) entity).getColor().name());
		}

		return key.toString().toLowerCase();
	}

	public String getMobDescription(Entity entity) {
		String key = getCapKey(entity);
		Location loc = entity.getLocation();
		return String.format("%s at (%s,%d,%d,%d)", key,
			loc.getWorld().getName(), loc.getBlockX(), loc.getBlockY(), loc.getBlockZ());
	}

	public boolean isSpecialMob(LivingEntity entity) {
		if (entity.getCustomName() != null) {
			return true;
		}

		if (entity instanceof Tameable) {
			Tameable tameable = (Tameable) entity;
			if (tameable.isTamed()) {
				return true;
			}
		}

		if (entity.getType() == EntityType.GUARDIAN && ((Guardian) entity).isElder()) {
			return true;
		}

		EntityEquipment equipment = entity.getEquipment();
		for (ItemStack armor : equipment.getArmorContents()) {
			if (armor != null && armor.getType() != Material.AIR) {
				return true;
			}
		}

		return false;
	}

	public void removeFromChunk(Chunk chunk) {
		Map<String, Integer> count = new HashMap<String, Integer>();

		for (Entity entity : chunk.getEntities()) {
			if (entity.isDead() || !(entity instanceof Animals || entity instanceof Monster)) {
				continue;
			}

			// Do not remove special mobs
			boolean special = isSpecialMob((LivingEntity) entity);
			if (special) {
				if (plugin.config.DEBUG) {
					plugin.getLogger().info("Special mob exempted from removal: " + getMobDescription(entity));
				}
				continue;
			}

			String key = getCapKey(entity);

			Integer oldCount = count.get(key);
			int mobCount = (oldCount == null) ? 1 : oldCount + 1;
			count.put(key, mobCount);

			int cap = getCap(key);
			if (cap >= 0 && mobCount > cap) {
				entity.remove();
				if (plugin.config.DEBUG) {
					plugin.getLogger().info("Removing mob " + getMobDescription(entity));
				}
			}
		}
	}

	public void removeFromAllChunks() {
		for (Chunk c : plugin.getServer().getWorlds().get(0).getLoadedChunks()) {
			removeFromChunk(c);
		}
	}

	public void applyAgeCap(Animals entity) {
		if (entity.getAgeLock() == true) {
			if (plugin.config.DEBUG) {
				plugin.getLogger().info("Age locked " + getMobDescription(entity) + " with age " + entity.getAge());
			}
			return;
		}

		if (plugin.config.AGE_CAP_BABY >= 0 && !entity.isAdult()) {
			entity.setAge(Math.max(entity.getAge(), -plugin.config.AGE_CAP_BABY));
		} else if (plugin.config.AGE_CAP_BREED >= 0 && entity.isAdult()) {
			entity.setAge(Math.min(entity.getAge(), plugin.config.AGE_CAP_BREED));
		}

		if (plugin.config.DEBUG) {
			plugin.getLogger().info("Age of " + getMobDescription(entity) + " capped to " + entity.getAge());
		}
	}

	public void limitCreatureSpawn(final CreatureSpawnEvent event) {
		CreatureSpawnEvent.SpawnReason reason = event.getSpawnReason();

		if ((reason == CreatureSpawnEvent.SpawnReason.BREEDING ||
			reason == CreatureSpawnEvent.SpawnReason.EGG ||
			reason == CreatureSpawnEvent.SpawnReason.DISPENSE_EGG) &&
			isFarmAnimal(event.getEntity())) {

			applyAgeCap((Animals) event.getEntity());

			for (Entity entity : event.getEntity().getNearbyEntities(4, 4, 4)) {
				if (isFarmAnimal(entity)) {
					applyAgeCap((Animals) entity);
				}
			}

			return;
		}

		boolean shouldLimitNaturalSpawn = plugin.config.LIMIT_NATURAL_SPAWN &&
			(event.getSpawnReason() == CreatureSpawnEvent.SpawnReason.BREEDING ||
				event.getSpawnReason() == CreatureSpawnEvent.SpawnReason.NATURAL ||
				event.getSpawnReason() == CreatureSpawnEvent.SpawnReason.DEFAULT);
		boolean shouldLimitSpawnerSpawn = plugin.config.LIMIT_SPAWNER_SPAWN &&
			event.getSpawnReason() == CreatureSpawnEvent.SpawnReason.SPAWNER;

		if (!shouldLimitNaturalSpawn &&
			!shouldLimitSpawnerSpawn) {
			return;
		}

		if (!plugin.config.SPAWN_LIMITED_ENTITY_TYPES.contains(event.getEntityType()) ||
			!hasCap(event.getEntity())) {
			return;
		}

		int cap = getCap((getCapKey(event.getEntity())));
		int count = 0;

		for (Entity otherEntity : event.getLocation().getChunk().getEntities()) {
			if (otherEntity.getType() == event.getEntityType()) {
				count++;
				if (count >= cap) {
					break;
				}
			}
		}

		if (count >= cap) {
			if (plugin.config.DEBUG) {
				plugin.getLogger().info("Cancel spawn of " + getMobDescription((event.getEntity())) +
					" (reason = " + event.getSpawnReason().name().toLowerCase() + ", cap = " + cap + ")");
			}
			event.getEntity().remove();
		}
	}

	private boolean isFarmAnimal(Entity entity) {
		return (entity instanceof Animals) && !(entity instanceof Tameable);
	}
}

File: src/main/kotlin/gg/obsidian/mobcontrol/MobManager.kt
package gg.obsidian.mobcontrol

import org.bukkit.Chunk
import org.bukkit.Material
import org.bukkit.entity.*
import org.bukkit.event.entity.CreatureSpawnEvent

import java.util.HashMap

class MobManager(val plugin: MobControl) {

  fun hasCap(entity: Entity): Boolean {
    return getCap(getCapKey(entity)) >= 0
  }

  fun getCap(key: String): Int {
    val cap = plugin.config.MOB_LIMITS[key]
    return if ((cap != null)) cap else -1
  }

  fun getCapKey(entity: Entity): String {
    val key = StringBuilder(entity.type.name)

    if (entity is Sheep) {
      key.append(entity.color.name)
    }

    return key.toString().toLowerCase()
  }

  fun getMobDescription(entity: Entity): String {
    val key = getCapKey(entity)
    val loc = entity.location
    return "%s at (%s,%d,%d,%d)".format(key, loc.world.name, loc.blockX, loc.blockY, loc.blockZ)
  }

  fun isSpecialMob(entity: LivingEntity): Boolean {
    if (entity.customName != null) {
      return true
    }

    if (entity is Tameable) {
      if (entity.isTamed) {
        return true
      }
    }

    if (entity.type == EntityType.GUARDIAN && (entity as Guardian).isElder) {
      return true
    }

    val equipment = entity.equipment
    for (armor in equipment.armorContents) {
      if (armor != null && armor.type != Material.AIR) {
        return true
      }
    }

    return false
  }

  fun removeFromChunk(chunk: Chunk) {
    val count = HashMap<String, Int>()

    for (entity in chunk.entities) {
      if (entity.isDead || !(entity is Animals || entity is Monster)) {
        continue
      }

      // Do not remove special mobs
      val special = isSpecialMob(entity as LivingEntity)
      if (special) {
        if (plugin.config.DEBUG) {
          plugin.logger.info("Special mob exempted from removal: " + getMobDescription(entity))
        }
        continue
      }

      val key = getCapKey(entity)

      val oldCount = count[key]
      val mobCount = if ((oldCount == null)) 1 else oldCount + 1
      count.put(key, mobCount)

      val cap = getCap(key)
      if (cap >= 0 && mobCount > cap) {
        entity.remove()
        if (plugin.config.DEBUG) {
          plugin.logger.info("Removing mob " + getMobDescription(entity))
        }
      }
    }
  }

  fun removeFromAllChunks() {
    for (c in plugin.server.worlds[0].loadedChunks) {
      removeFromChunk(c)
    }
  }

  fun applyAgeCap(entity: Animals) {
    if (entity.ageLock == true) {
      if (plugin.config.DEBUG) {
        plugin.logger.info("Age locked " + getMobDescription(entity) + " with age " + entity.age)
      }
      return
    }

    if (plugin.config.AGE_CAP_BABY >= 0 && !entity.isAdult) {
      entity.age = Math.max(entity.age, -plugin.config.AGE_CAP_BABY)
    } else if (plugin.config.AGE_CAP_BREED >= 0 && entity.isAdult) {
      entity.age = Math.min(entity.age, plugin.config.AGE_CAP_BREED)
    }

    if (plugin.config.DEBUG) {
      plugin.logger.info("Age of " + getMobDescription(entity) + " capped to " + entity.age)
    }
  }

  fun limitCreatureSpawn(event: CreatureSpawnEvent) {
    val reason = event.spawnReason

    if ((reason == CreatureSpawnEvent.SpawnReason.BREEDING || reason == CreatureSpawnEvent.SpawnReason.EGG || reason == CreatureSpawnEvent.SpawnReason.DISPENSE_EGG) && isFarmAnimal(event.entity)) {

      applyAgeCap(event.entity as Animals)

      for (entity in event.entity.getNearbyEntities(4.0, 4.0, 4.0)) {
        if (isFarmAnimal(entity)) {
          applyAgeCap(entity as Animals)
        }
      }

      return
    }

    val shouldLimitNaturalSpawn = plugin.config.LIMIT_NATURAL_SPAWN && (event.spawnReason == CreatureSpawnEvent.SpawnReason.BREEDING || event.spawnReason == CreatureSpawnEvent.SpawnReason.NATURAL || event.spawnReason == CreatureSpawnEvent.SpawnReason.DEFAULT)
    val shouldLimitSpawnerSpawn = plugin.config.LIMIT_SPAWNER_SPAWN && event.spawnReason == CreatureSpawnEvent.SpawnReason.SPAWNER

    if (!shouldLimitNaturalSpawn && !shouldLimitSpawnerSpawn) {
      return
    }

    if (!plugin.config.SPAWN_LIMITED_ENTITY_TYPES.contains(event.entityType) || !hasCap(event.entity)) {
      return
    }

    val cap = getCap((getCapKey(event.entity)))
    var count = 0

    for (otherEntity in event.location.chunk.entities) {
      if (otherEntity.type == event.entityType) {
        count++
        if (count >= cap) {
          break
        }
      }
    }

    if (count >= cap) {
      if (plugin.config.DEBUG) {
        plugin.logger.info("Cancel spawn of " + getMobDescription((event.entity)) + " (reason = " + event.spawnReason.name.toLowerCase() + ", cap = " + cap + ")")
      }
      event.entity.remove()
    }
  }

  private fun isFarmAnimal(entity: Entity): Boolean {
    return (entity is Animals) && entity !is Tameable
  }
}
